### DISCUSSION TITLE
How to seach for certain Go projects that has a main func?

### DISCUSSION THREAD

**Discussion by tigerinus** [community/community#28664] [Other Feature Feedback, Questions, & Ideas] [answered]
Hello,

I want to search for some Golang projects and one of the conditions is the project must have a `main` func.

Any hint please?

Thanks!

**Comment by BergerAPI** [comment]
You could use [Github Codesearch](https://cs.github.com/). With the following query it will look for go projects that match with `func main(`.
```
language:go "func main("
```

The direct link to the search query is [this](https://cs.github.com/?scopeName=All+repos&scope=&q=language%3Ago+%22func+main%28%22).

**Reply by tigerinus** [reply]
This is great.

Would it be possible to return just the repo owner/name that satisfies this search condition?

**Reply by BergerAPI** [reply]
I haven't found any official API-Documentations from Github, and looking at the Network Tab also results in nothing. This means you will need to parse the HTML and get the results yourself. You can do this with a library [cheerio.js](https://cheerio.js.org/) for example.

**Reply by look** [reply]
@tigerinus there is not a way to return only the repository at this time. An API is on our roadmap for the new code search, but it's not ready yet.

**Comment by BergerAPI** [comment, marked as answer]
There is a Github-GraphQL-API which allows you to search for this and gather the results as array.
The query to achieve what you are looking for looks like this:

```graphql
{
  search(query: "language:go func main", type: REPOSITORY, first: 10) {
    repos: edges {
      repo: node {
        ... on Repository {
          url
          name
          createdAt
          updatedAt
        }
      }
    }
  }
}
```

It will respond with 10 repos and their url, name, creation date and update date. 
Looks like this:

```json
{
  "data": {
    "search": {
      "repos": [
        {
          "repo": {
            "url": "https://github.com/ShubhamSupekar/Bubble-Sorting-Algorithm",
            "name": "Bubble-Sorting-Algorithm",
            "createdAt": "2021-12-17T10:25:55Z",
            "updatedAt": "2021-12-19T13:36:25Z"
          }
        }
      ]
    }
  }
}
```

You can send this request like this:

## Axios

```js
const response = await axios.post(
        'https://api.github.com/graphql',
        {
            query: '{' +
                '  search(query: "language:go func main", type: REPOSITORY, first: 1) {' +
                '    repos: edges {' +
                '      repo: node {' +
                '        ... on Repository {' +
                '          url' +
                '          name' +
                '          createdAt' +
                '          updatedAt' +
                '        }' +
                '      }' +
                '    }' +
                '  }' +
                '}',
        },
        {
            headers: {
                'Authorization': 'bearer token',
                'Content-Type': 'application/x-www-form-urlencoded'
            }
        }
    );

console.log(JSON.stringify(response.data, null, 4))
```

# Authorization

You will need to set an Auth-Header in the request. The format for this looks like this:
``Authorization: bearer token``. You can get this token like described [here](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token)

**Comment by alexalannunes** [comment]
I'm very excited to use https://cs.github.com/about
